File: hs/Day10.hs
#!runghc
{-# OPTIONS_GHC -Wall #-}

module Main where

import Data.List
import Data.Map (Map)
import qualified Data.Map as M
import Data.Maybe

readInputs :: IO [Int]
readInputs = sort . map read . lines <$> readFile "../inputs/Day10.txt"

reachable :: [Int] -> Int -> [Int]
reachable adapters pos = reachable' pos
  where
    size = length adapters
    joltage = adapters !! pos
    reachable' cur
      | cur == size = []
      | otherwise =
        if adapters !! cur - joltage <= 3
          then cur : reachable' (cur + 1)
          else []

ways :: [Int] -> Int
ways adapters = fromJust $ M.lookup 0 $ foldl f initialMemo reversedIndices
  where
    size = length adapters
    initialMemo = M.fromList [(size - 1, 1)]
    indices = [0 .. size - 1]
    reversedIndices = reverse indices
    f :: Map Int Int -> Int -> Map Int Int
    f memo pos = M.insert pos res memo
      where
        reachableFromPos = reachable adapters pos
        res = sum $ map readOrZero reachableFromPos
        readOrZero = fromMaybe 0 . flip M.lookup memo

main :: IO ()
main = do
  adapters <- readInputs
  let pairs = mkPairs adapters
  let diff k = 1 + length (filter (\(a, b) -> b - a == k) pairs)
  print $ diff 1
  print $ diff 3
  print $ diff 3 * diff 1
  print $ ways (0 : adapters)
  where
    mkPairs x = zip x (tail x)

File: py/day10.py
#!python3

from queue import Queue

lines = open("../inputs/Day10.txt", "r").readlines()

# Problem 1

adapters = sorted([int(line) for line in lines])

counts = {1: 1, 3: 1}

for (a, b) in zip(adapters, adapters[1:]):
	counts[b - a] +=  1

print(counts[1] * counts[3])

adapters.insert(0, 0)

from collections import defaultdict

ways = defaultdict(lambda: 0)
ways[len(adapters) - 1] = 1

for cur in reversed(range(0, len(adapters))):

	try:
		if 1 <= adapters[cur + 1] - adapters[cur] <= 3:
			ways[cur] += ways[cur + 1]
	except IndexError:
		...

	try:
		if 1 <= adapters[cur + 2] - adapters[cur] <= 3:
			ways[cur] += ways[cur + 2]
	except IndexError:
		...

	try:
		if 1 <= adapters[cur + 3] - adapters[cur] <= 3:
			ways[cur] += ways[cur + 3]
	except IndexError:
			...

print(ways[0])